chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -130,16 +130,30 @@ pub fn pre_validate_chunk_state_witness(
     // Blocks from the last last new chunk (exclusive) to the last new chunk (inclusive).
     let mut blocks_after_last_last_chunk = Vec::new();
 
-    {
+    let (last_chunk_shard_id, last_chunk_shard_index) = {
         let mut block_hash = *state_witness.chunk_header.prev_block_hash();
         let mut prev_chunks_seen = 0;
+
+        let epoch_id = state_witness.epoch_id;
+        let shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+
+        // It is ok to use the shard id from the header because it is a new
+        // chunk. An old chunk may have the shard id from the parent shard.
+        let mut current_shard_id = state_witness.chunk_header.shard_id();
+        let mut current_shard_index = shard_layout.get_shard_index(current_shard_id);
+
+        let mut last_chunk_shard_id = current_shard_id;
+        let mut last_chunk_shard_index = current_shard_index;
         loop {
             let block = store.get_block(&block_hash)?;
+            (current_shard_id, current_shard_index) =
+                epoch_manager.get_prev_shard_id(&block_hash, current_shard_id)?;
+
             let chunks = block.chunks();
-            let Some(chunk) = chunks.get(shard_index) else {
+            let Some(chunk) = chunks.get(current_shard_index) else {
```

Comment:
Is that off by one block? 
the shard_id and shard_index are from the prev block (that's what get_prev_shard_id gives)
the block.chunks() are from the current block

Reply:
I think it was correct. However, I made it more intuitive - by changing `current_shard_id` simultaneously with going to prev block.